ACPs/113-provable-randomness/README.md

```diff
@@ -0,0 +1,157 @@
+```text
+ACP: 113
+Title: Provable Virtual Machine Randomness
+Author(s): Tsachi Herman <http://github.com/tsachiherman>
+Discussions-To: <GitHub Discussion URL (POPULATED BY MAINTAINER, DO NOT SET)>
+Status: Proposed
+Track: Standards
+Replaces (*optional): n/a
+Superseded-By (*optional): n/a
+```
+
+## Abstract
+
+Avalanche offers developers flexibility through subnets and EVM-compatible smart contracts. However, the platform's deterministic block execution limits the use of traditional random number generators within these contracts.
+
+To address this, a mechanism is proposed to generate verifiable, non-cryptographic random number seeds on the Avalanche platform. This method ensures uniformity while allowing developers to build more versatile applications.
+
+
+## Motivation
+
+Reliable randomness is essential for building exciting applications on Avalanche. Games, participant selection, dynamic content, supply chain management, and decentralized services all rely on unpredictable outcomes to function fairly. Randomness also fuels functionalities like unique identifiers and simulations. Without a secure way to generate random numbers within smart contracts, Avalanche applications become limited.
+
+Avalanche's traditional reliance on external oracles for randomness creates complexity and bottlenecks. These oracles inflate costs, hinder transaction speed, and are cumbersome to integrate. As Avalanche scales to more Subnets, this dependence on external systems becomes increasingly unsustainable.
+
+A solution for verifiable random number generation within Avalanche solves these problems. It provides fair randomness functionality across the chains, at no additional cost. This paves the way for a more efficient Avalanche ecosystem.
+
+## Specification
+
+### Changes Summary
+
+Existing avalanche protocol breaks the block building into two parts : external and internal. The external block is the SnowMan++ block, whereas the internal block is the actual virtual machine block.
+
+To support randomness, a BLS based VRF implementation is used, that would be recursively signing its own signatures as its message. Since the BLS signatures are deterministic, they provide a great way to construct a reliable VRF.
+
+For proposers that do not have a BLS key associated with their node, the hash of the signature from the previous round.
+
+In order to bootstrap the signatures chain, a missing signature would be replaced with a byte slice that is the hash product of a verifiable and trustable seed.
+
+The changes proposed here would affect the way a block is being validated. Therefore, when this change gets implemented, it needs to be deployed as a mandatory upgrade.
```

Comment:
> it needs to be deployed as a mandatory upgrade.


Is this feature always activated, or would it be possible to opt-in to activate this feature via some kind of voting?
It might be that some subnet/L1 would prefer to not pay the extra overhead.

Reply:
A subnet would still be fully usable without having this feature active. From technical perspective, the subnet would need to be configured to set the activation time for this feature to infinity. I think that the technical aspects (how to enabled/disable) of this feature are a bit beyond the scope of this ACP.